### ISSUE TITLE
Theme: Login page customization

### ISSUE THREAD

**Issue by swelljoe** [qooob/authentic-theme#639]
See here for some discussion: https://www.virtualmin.com/node/44933

I think the ability to replace the Webmin/Usermin logo on login and customize the text on the login page is very reasonable. (Localization makes that somewhat more complex. Not sure how we'd handle localization of user input data, but I guess most users would only want one language, and that'd be the language they'd used for their text without needing any decisions from the software.)

**Comment by qooob** [comment]
Hi, Joe.

You don't need anything at the moment to make your login screen look completely different.

Theme is fully `hackable`. All you need is to use theme extensions editor for _CSS_.

This is how you could make Webmin Logo and its text from not showing:

```CSS
.session_login form[method="post"] > i,
.session_login form[method="post"] > i + h2 {
	display:none;
}
```

You can even change the look of login screen utterly. (I noticed that custom scripts (extensions: js) are not embedded for login page. That will be fixed.)

In the future, I hope, I will be able to find time to make selectable skins for login page.

**Comment by qooob** [comment]
..and one more.

If you want to override the default text labels, first of all - find the desired key in lang file of the module/theme.

Let's say that you want to change word <em>Webmin</em> on login page of **Authentic Theme** from <em>Webmin</em> to <em>Supermin</em>, then we take <em>theme_xhred_titles_wm=Webmin</em> from <em>/usr/(libexec|share)/webmin/authentic-theme/lang/en.UTF8</em>.

Later, create <em>custom-lang</em> file in configuration directory <em>/etc/webmin/authentic-theme</em> of the theme. (or other module if you want to tweak that)

To see what I mean just run on the console the following:

```BASH
echo "theme_xhred_titles_wm=Supermin" > /etc/webmin/authentic-theme/custom-lang
```

This kind of customization will survive updates and you would need to do it only once.

Good luck ;)

**Comment by qooob** [comment]
Now it's possible to use custom js scripts for login page as well.

**Comment by brianforan** [comment]
@qooob to edit the usermin logo is it the same process as editing the webmin logo? I've been poking around online but most forum posts are from around 2009 with similar questions. I was looking for `/etc/usermin` but that does not exist so I'm not entirely sure where I should be looking.

**Comment by qooob** [comment]
You should be looking into `/etc/usermin`.

You could just copy `/etc/webmin` into `/etc/usermin`.

It will work.

**Comment by qooob** [comment]
Actually, I meant to say:

`/etc/webmin/authentic-theme` into `/etc/usermin/authentic-theme`

**Comment by qooob** [comment]
At them moment the logo for Webmin and Usermin is the same.

If you want to have separate logo for Usermin just upload those to `/etc/usermin/authentic-theme`:

`logo_welcome.png`

and 

`logo.png`

**Comment by brianforan** [comment]
And those won't be overwritten by updates if I use a custom company logo? That was my primary concern, it seemed like a lot of things that were edited may be overwritten by updates. Like the custom-lang file for example, I couldn't find any documentation or anything on where that came from, so I never tried anything.

**Comment by qooob** [comment]
Updates don't overwrite configurations.

In your case, for now just store logos for Usermin somewhere safe. Because if you upload new logos in Webmin they will be overwritten.

Please open the separate issue to provide ability to have different logos for Webmin/Usermin if that's an issue.

In this post above you can find an [explanation](https://github.com/qooob/authentic-theme/issues/639#issuecomment-271138852) of how to use custom language file.

**Comment by qooob** [comment]
Theme Extension Editor already contains custom-lang file support.

All you need to do is to find string that you want to be displayed differently.

Search for strings you want to change in `/lang` folder.

**Comment by NanoTest5000** [comment]
Hi iliarostovtsev... I have been enjoying Authentic-Theme.... Its really awesome

I would like to know how to customize **Virtual Login Text** ( You must enter a username and password) to something else....

**Comment by iliarostovtsev** [comment]
> I would like to know how to customize Virtual Login Text ( You must enter a username and password) to something else....

Please could you clarify what you mean, and/or attach a screenshot?

**Comment by NanoTest5000** [comment]
[image: Demo Login.JPG]
*I want to customize the Text in the login Page... I couldn't locate the
Source file of the login Page...*

*Please let me know how to change it....  I want Webmin Text and below
lines to be customized...*

**Comment by NanoTest5000** [comment]
Ok I have shared the link of the image

https://drive.google.com/file/d/1ENdKDKQDbbbZE3cscnJMcx7wgZU-mrei/view?usp=sharing

[image: Demo Login.JPG]
*I want to customize the Text in the login Page... I couldn't locate the
Source file of the login Page...*

*Please let me know how to change it....  I want Webmin Text and below
lines to be customized...*

*Thanks a lot*

**Comment by iliarostovtsev** [comment]
> I would like to know how to customize Virtual Login Text ( You must enter a username and password) to something else

You would have to create `/etc/webmin/authentic-theme/custom-lang` file and put there the following:

```
login_message=This server only supports blood-sample authentication, please touch the fingerprint scanner with bleeding index finger of your right hand, and wait patiently. Left hand authentication, just as authentication using other fingers is not yet supported. Sorry!
```

.. other language strings that can be replaced can be [found here](https://github.com/authentic-theme/authentic-theme/blob/master/lang/en#L49).

**Comment by NanoTest5000** [comment]
*Thanks For the Fast Reply.... U r really Awesome... I have been searching
for two more things...*

*One is to change the Webmin Text and Logo in Dashboard.*

*And Second is*

*How to Hide Tab Title or Customize Tab Title to Show the Company Name*

*Please Help me out....*

*I have been searching more than a week couldn't find a solution...*

*That's y i contacted u.... Never expected u would reply fast....*

*Excellent Support*

**Comment by iliarostovtsev** [comment]
> *One is to change the Webmin Text and Logo in Dashboard.*

It is always best to provide a screenshot, to make sure we're on the same page. You can copy/paste screenshots directly to GitHub editor window, no need to upload elsewhere.

> *How to Hide Tab Title or Customize Tab Title to Show the Company Name*

The same, I have no idea what you're talking about and I won't dare to guess. 😉

**Comment by iliarostovtsev** [comment]
> How to Change Webmin Text

The text should be changed the same way as described in the comment above for `login_message` but this time using `theme_xhred_titles_wm` key.

> How to Change the Logo

This can only be done using CSS theme extensions by replacing a content of an image with needed content from [default icon pack](https://authentic-theme.com/kit/) or by placing own icon pack in the CSS theme extension file.

Example:

~~~CSS
.wbm-webmin:before
{
    content: '\f00a';
}
~~~

Technically you could also use theme overlay to load CSS but there is no real need for doing this, as you can already do it using Authentic Theme extensions editor, like this:

![image](https://user-images.githubusercontent.com/4426533/111524158-27920c80-876d-11eb-88db-e238f7ec67eb.png)


> How to customize tab title to show

This will only be supported starting with upcoming Authentic Theme 1.974 and will depend on modifying already mentioned  `theme_xhred_titles_wm` using `custom-lang` file.

**Comment by NanoTest5000** [comment]
**_Thanks iliarostovtsev...._** Everything Worked Great... Waiting for Authentic Theme 1.974... When is next update???

**Comment by iliarostovtsev** [comment]
Most likely tomorrow.

**Comment by NanoTest5000** [comment]
**_Hi iliarostovtsev_**   I'm on **_19.74-RC3_**... How to Change **Page Title** (Dashboard to Show the Company Name)... And Proper way to change the **Favicon**
![Dashboard Title](https://user-images.githubusercontent.com/80373260/111894629-72907600-8a32-11eb-84d4-59acdd6ee96d.JPG)

**Comment by iliarostovtsev** [comment]
> How to Change Page Title (Dashboard to Show the Company Name or hide it completely).

This is already supported for checked devel version of the theme (and described above how to change language keys). The rest is configurable from Theme Configuration page - 

![image](https://user-images.githubusercontent.com/4426533/112048854-e41a1280-8b5f-11eb-93de-b818eea34439.png)

> And Proper way to change the Favicon

Not yet possible. But I think I could provide support for this. Is this really needed?

> Waiting for Authentic Theme 19.74... When is next update???

Working on new features. Delayed a bit. Coming soon..

**Comment by NanoTest5000** [comment]
**_Thanks a Lot @iliarostovtsev_** I was waiting for ur reply... I'm really addicted to authentic-theme.... **It's Really Awesome**....

**Thanks a Lot for this Beautiful Theme**.....



> **Not yet possible. But I think I could provide support for this. Is this really needed?**

**Yes Please Provide support for this.... I really need it... I want to have custom Favicon**

**Comment by iliarostovtsev** [comment]
> Yes Please Provide support for this.... I really need it... I want to have custom Favicon

Okay, I will see what I can do.

**Comment by NanoTest5000** [comment]
> **Okay, I will see what I can do.**

**Ok I'll be waiting....**

**Comment by iliarostovtsev** [comment]
Okay, I have added ability to customize _favicons_ as well. All you need to do is to copy to `/etc/webmin/authentic-theme/favicons` the modified content, which should be based on initial content found in original package under `authentic-theme/images/favicons` directory.

----

Next, another big change is coming with **19.74-RC4**, which is ability to utilize on-load brandings. It is familiar to what Google does in _**Gmail**_. By default all users will see the following branded splash screen, based on the product they are currently using:

![image](https://user-images.githubusercontent.com/4426533/112234396-819f4000-8c4d-11eb-902e-555b9bde16a7.png) ![image](https://user-images.githubusercontent.com/4426533/112234408-895ee480-8c4d-11eb-9ecc-33fa072e8787.png) ![image](https://user-images.githubusercontent.com/4426533/112234434-97ad0080-8c4d-11eb-912b-530d3beacd74.png) ![image](https://user-images.githubusercontent.com/4426533/112234448-a1ceff00-8c4d-11eb-84db-fbb886447081.png)

This however can be disabled using theme configuration page and its new options: 

![image](https://user-images.githubusercontent.com/4426533/112235730-49e5c780-8c50-11eb-8c03-c4a01840c562.png)


What is **important to know** about this new feature besides that it can be disabled:

  1. Branding splash screen will be shown only once on full page reload
  2. Branding splash screen can be dismissed manually with hitting _Esc_ two times or double clicking with a mouse (after min timeout elapsed)
  3. Branding splash screen can be customized using the same principal described for _favicons_ but using `brand` directory
  4. Branding splash screen can be customized (additionally, not needed in most cases) using using `custom_embed_product_branding` sub defined in `scripts.pl` under theme Perl extensions
  5. Internally, branding splash screen loader is _pretty smart_, and will be shown for at least 1 second or at max for 6 seconds. If the initial page load happens before default minimum 1 second timeout, the branding will still be shown. However, if the page load takes more than 1 second and less than 6, than the branding splash screen will wait for the page to load fully and be automatically dismissed. If the page load takes more than 6 seconds (due to an error or just a slow server), then branding splash screen will be dismissed anyway. The amount of min and max seconds for the splash screen to show is configurable. Although, not in UI but by using a browser's console with `settings_product_branding_show_time_min_privileged` and `settings_product_branding_show_time_max_privileged` options, which can be saved with `theme.config.save()` command after tweaking. The sanity check is done for min is 1 to 10 seconds and for max is 1 to 120 seconds.

**Comment by NanoTest5000** [comment]
**_Thanks a Lot @iliarostovtsev_**

**Comment by iliarostovtsev** [comment]
I wish we could use a single svg for favicon, tiny and nice. :slightly_frowning_face:

**Comment by NanoTest5000** [comment]
**Hi @iliarostovtsev I'm Really Sorry for disturbing u... But i really need ur help on my Project...**

I know we can change the titles like **theme_xhred_titles_wm** in **custom-lang file** 

But There's no option to change the **Title** in **Login Page of Webmin and Login Page of Usermin**

I want to **Customize** the **Login to Webmin** and **Login to Usermin** to Something Else

**Please Help me I really couldn't figure out...**

I have attached the Screenshot

![Login Page](https://user-images.githubusercontent.com/80373260/113137959-d2f99180-9242-11eb-94dd-3a686a111212.JPG)

**Comment by iliarostovtsev** [comment]
> I want to Customize the Login to Webmin and Login to Usermin to Something Else


Check:

```
/usr/{libexec|share}/webmin/lang/en
```

.. and its `session_header` key - later use custom `/etc/webmin/custom-lang` file to override it.

---

Do you really own? :slightly_smiling_face:

**Comment by NanoTest5000** [comment]
**Thank u soo much @iliarostovtsev** ❤

> Do you really own cloud.com domain? 🙂 

I own many domain's but cloud.com is not mine 😉

**Comment by NanoTest5000** [comment]
**Hi @iliarostovtsev**   Is it Possible to disable the favicon ???

**Comment by iliarostovtsev** [comment]
> How to subdomain wildcard URL redirect

Not sure what you're asking. Redirect from where to where (what is the full URL before and expected after the redirect)?

> Is it Possible to disable the favicon

By creating empty pngs for custom favcons described above.

**Comment by iliarostovtsev** [comment]
> Is it Possible to disable the favicon ???

Check this commit out - https://github.com/authentic-theme/authentic-theme/commit/4019e58a4e312d3eda80fe76c66cf968093b4728

**Comment by iliarostovtsev** [comment]
> Check this commit out - 4019e58

Actually this one - https://github.com/authentic-theme/authentic-theme/commit/40f43586eddf33ae8541412a10afb712b2040b77

**Comment by NanoTest5000** [comment]
> Not sure what you're asking. Redirect from where to where (what is the full URL before and expected after the redirect)?

anything.domain.com to domain.com

or  

anything.domain.com to https://domain.com

I want all subdomains like mail.domain.com and whatever.domain.com  to go to my main domain ( domain.com or https://domain.com )

i tried using wildcard * But it didn't work

**Comment by iliarostovtsev** [comment]
> I want all subdomains like mail.domain.com and whatever.domain.com to go to my main domain ( domain.com or https://domain.com )

Cannot help with that. I suggest you check for webserver documentation. Setting up redirects is quite common question.
